### ISSUE TITLE
Color problems with semitransparency & mipmapping enabled since ~5.4.0

### ISSUE THREAD

**Issue by jeremyshannon** [minetest/minetest#15401]
### Luanti version

```true
5.4.0 - 5.10.0 RC1
```

### Irrlicht device

X11, SDL

### Operating system and version

Devuan Daedalus (eq to Debian Bookworm)

### GPU model

AMD RX 480

### Summary

Transparency changes in 5.4.0 and 5.5.0 caused a couple of visual regressions. Here's a screenshot of Exile in MT 5.3.0, looking as intended:
![530](https://github.com/user-attachments/assets/059b8439-5cfd-475e-ab8e-d767095c116a)

Here's the same scene in 5.5.0, where the window is grossly miscolored:
![550](https://github.com/user-attachments/assets/f5a890f3-0455-4f59-b92c-757a7f0a95b0)

All following versions display the same behavior. The window coloring problem only occurs when mipmapping is enabled. Turning it off prevents this problem in current Minetest.

(I switched off mipmaps way back when over this, but recently decided to try again, and then to track down when this got busted. I made these screenshots a while back, I thought I had the renderer recorded but I don't see that here, I believe it's openGL 4.6 in all of these.
 I can also bisect things if that's needed.)

### Steps to reproduce

Enable mipmapping, look at a clear glass window in Exile

**Comment by sfan5** [comment]
I got a crash on first launch:
```
2024-11-09 00:13:15: ERROR[Main]: ServerError: AsyncErr: Lua: Runtime error from mod 'lore' in callback on_playerReceiveFields(): .../games/exile/mods/lore/login.lua:116: attempt to call field 'mt_required_version' (a nil value)
```

Anyway the moon glass is indeed not semi-transparent.

A look at the definition reveals this:
```lua
minetest.register_node("artifacts:moon_glass", {
        description = "Moon Glass",
        drawtype = "glasslike",
        tiles = {"artifacts_moon_glass.png"},
        stack_max = minimal.stack_max_bulky *4,
        light_source = 5,
        paramtype = "light",
        sunlight_propagates  = true,
        use_texture_alpha = c_alpha.clip,
        sounds = nodes_nature.node_sound_glass_defaults(),
        groups = {cracky = 1,},
        after_place_node = minimal.protection_after_place_node,
})
```
Texture alpha "clip" does not support semi-transparency. Refer to lua_api.md.

**Comment by appgurueu** [comment]
> Texture alpha "clip" does not support semi-transparency. Refer to lua_api.md.

This explains the "minor" transparency problems reported here, but not the mipmapping issue:

> and now for the big one: Here's the same scene in 5.5.0, where the window is grossly miscolored

**Comment by jeremyshannon** [comment]
Yes, the window thing with mipmapping enabled is what I was actually concerned about.

(Good to know the moonglass transparency problems are a bug in Exile, that was something I only noticed when I was taking the screenshots for the window bug. I'll have to audit our nodes to see if any others got the wrong setting)
 I'm also surprised to find the stable release was crashing with a new player! I missed that when testing it, but it's fixed now.)

**Comment by sfan5** [comment]
its the `imageCleanTransparent` algorithm doing this, simple test:
```diff
diff --git a/src/client/imagefilters.cpp b/src/client/imagefilters.cpp
--- a/src/client/imagefilters.cpp
+++ b/src/client/imagefilters.cpp
@@ -161,6 +161,7 @@ static void imageCleanTransparentWithInlining(video::IImage *src, u32 threshold)
  */
 void imageCleanTransparent(video::IImage *src, u32 threshold)
 {
+       return;
        if (src->getColorFormat() == video::ECF_A8R8G8B8)
                imageCleanTransparentWithInlining<true>(src, threshold);
        else
```

It's no surprise that it messes up the texture, since it shouldn't be applied to semi-transparent textures in the first place.

**Comment by sfan5** [comment]
Fixing this is not trivial since inside `TextureSource::getTextureForMesh` we do not know if the texture is about to be used for semi-transparent or clip/opaque rendering.
But we can go and give it this information.

**Comment by Desour** [comment]
Is there a reason why we need to set the colors of semi-transparent texture pixels, aka why we call `imageCleanTransparent` with threshold 127? If it's semi-transparent, the modder has set the color explicitly anyways, and image size optimizers don't discard the color.

```diff
diff --git a/src/client/imagesource.cpp b/src/client/imagesource.cpp
index 0f4e81f97..1c620feff 100644
--- a/src/client/imagesource.cpp
+++ b/src/client/imagesource.cpp
@@ -1514,7 +1514,7 @@ bool ImageSource::generateImagePart(std::string_view part_of_name,
 			// Apply the "clean transparent" filter, if needed
 			if (m_setting_mipmap || m_setting_bilinear_filter ||
 				m_setting_trilinear_filter || m_setting_anisotropic_filter)
-				imageCleanTransparent(baseimg, 127);
+				imageCleanTransparent(baseimg, 0);
 
 			/* Upscale textures to user's requested minimum size.  This is a trick to make
 			 * filters look as good on low-res textures as on high-res ones, by making

```

----

Also, information for other readers, because I was confused:
* Issue is no longer about moonglass (the grey cube on the left), but about `tech:window_clear` (the panes with the brown borders).
* `window_clear` uses blend=alpha, and this texture: https://codeberg.org/Mantar/Exile/src/branch/master/mods/tech/textures/tech_clear_glass_window.png

**Comment by sfan5** [comment]
> Is there a reason why we need to set the colors of semi-transparent texture pixels, aka why we call `imageCleanTransparent` with threshold 127?

the function comment explains it:
```
 * Parameter "threshold" is the alpha level below which pixels are considered
 * transparent. Should be 127 when the texture is used with ALPHA_CHANNEL_REF,
 * 0 when alpha blending is used.
```
in theory the texture of ALPHA_CHANNEL_REF material could contain junk data in pixels with alpha <127 and the engine should handle it correctly.
in practice nobody would do this.

we could consider just applying this fix. the other alternative is differentiating between the used alpha type at `getTextureForMesh()` call time, which is more work.

**Comment by Wuzzy2** [comment]
I believe I bumped into this bug for the white glass in Eyeballs v1.0.2. This game uses 32×32 textures for the glass.

This is the Eyeballs issue (with screenshot): https://codeberg.org/Wuzzy/Eyeballs/issues/17

This bug only manifests if one of mipmapping, bi-/tri-/anisotropic filtering is enabled.

In the meanwhile, I already applied a workaround for my game. If the game detects one of these settings as active (this is possible because it's a singleplayer game), it will use a simplified fallback texture that only uses one color.
